PubNub/Data/Models/PNPAMToken.h

```diff
@@ -0,0 +1,161 @@
+
+#pragma mark - Interfaces declaration
+
+/**
+ * @brief \b PubNub Access Manager resource permission.
+ *
+ * @discussion Object represent single resource permission information.
+ *
+ * @author Serhii Mamontov
+ * @version 4.17.0
+ * @since 4.17.0
+ * @copyright © 2010-2021 PubNub, Inc.
+ */
+@interface PNPAMResourcePermission : NSObject
+
+
+#pragma mark - Information
+
+/**
+  * @brief Value with which instance has been initialized.
+ */
+@property (nonatomic, readonly, assign) PNPAMPermission value;
+
+
+#pragma mark -
+
+
+@end
+
+
+/**
+ * @brief \b PubNub Access Manager resource permissions.
+ *
+ * @discussion Object represent resource permissions stored in PAM token.
+ *
+ * @author Serhii Mamontov
+ * @version 4.17.0
+ * @since 4.17.0
+ * @copyright © 2010-2021 PubNub, Inc.
+ */
+@interface PNPAMTokenResource : NSObject
+
+
+#pragma mark - Information
+
+/**
+ * @brief Permissions granted to specific / regexp matching channels.
+ */
+@property (nonatomic, readonly, strong) NSDictionary<NSString *, PNPAMResourcePermission *> *channels;
+
+/**
+ * @brief Permissions granted to specific / regexp matching channel groups.
+ */
+@property (nonatomic, readonly, strong) NSDictionary<NSString *, PNPAMResourcePermission *> *groups;
+
+/**
+ * @brief Permissions granted to specific / regexp matching uuids.
+ */
+@property (nonatomic, readonly, strong) NSDictionary<NSString *, PNPAMResourcePermission *> *uuids;
+
+#pragma mark -
+
+
+@end
+
+
+/**
+ * @brief \b PubNub Access Manager Token.
+ *
+ * @discussion Object represent decoded token content.
+ *
+ * @author Serhii Mamontov
+ * @version 4.17.0
+ * @since 4.17.0
+ * @copyright © 2010-2021 PubNub, Inc.
+ */
+@interface PNPAMToken : NSObject
+
+
+#pragma mark - Information
+
+/**
+ * @brief Token version.
+ */
+@property (nonatomic, readonly, assign) NSUInteger version;
+
+/**
+ * @brief Token generation date time.
+ */
+@property (nonatomic, readonly, assign) NSUInteger timestamp;
+
+/**
+ * @brief Maximum duration (in minutes) during which token will be valid.
+ */
+@property (nonatomic, readonly, assign) NSUInteger ttl;
+
+/**
+ * @brief The uuid that is exclusively authorized to use this token to make API requests.
+ */
+@property (nonatomic, nullable, readonly, strong) NSString *authorizedUUID;
+
+/**
+ * @brief Permissions granted to specific resources.
+ */
+@property (nonatomic, readonly, strong) PNPAMTokenResource *resources;
+
+/**
+ * @brief Permissions granted to resources which match specified regular expression.
+ */
+@property (nonatomic, readonly, strong) PNPAMTokenResource *patterns;
+
+/**
+ * @brief Additional information which has been added to the token.
+ */
+@property (nonatomic, readonly, strong) NSDictionary *meta;
+
+/**
+ * @brief PAM token content signature.
+ */
+@property (nonatomic, readonly, strong) NSData *signature;
+
+/**
+ * @brief Whether provided PAM token string was valid and properly processed.
+ */
+@property (nonatomic, readonly, assign) BOOL valid;
```

Comment:
Do we need this? I usually tend to limit the state variables. Looks like it can be derived from the presence of error which is devined below.

Reply:
@bartk maybe it should be computed property and check `ttl` to figure out, whether token expired or not.
But same can be done by setting corresponding error after token parsed. Last approach maybe better :)

Reply:
👍